File: src/utils/color.py
"""File containing the color class for a disease color in the game of Pandemic."""
class Color:
    def __init__(self, color: str):
        """Initialize a color with a name."""
        if color in ["blue", "yellow", "black", "red"]:
            self._name = color
        else:
            raise ValueError(f"Invalid color: {color}\n Valid colors are: blue, yellow, black, red.")
        self.disease_cubes = 24
        self.cured = False
        self.eradicated = False

    @property
    def name(self) -> str:
        return self._name

    def __str__(self) -> str:
        """Return the color of the disease."""
        return repr(self)
    
    def __repr__(self) -> str:
        """Return a representation of the color."""
        return f"Color(color='{self.name}', disease_cubes={self.disease_cubes}, cured={self.cured}, eradicated={self.eradicated})"
    
    def __hash__(self) -> int:
        """Return the hash of the color."""
        return hash(self.name)
    
    def cure(self) -> None:
        """Cure the disease."""
        self.cured = True

    def eradicate(self) -> None:
        """Eradicate the disease."""
        if self.disease_cubes == 24 and self.cured:
            self.eradicated = True

File: cpp/Private/utils/color.cpp
namespace pandemic {
	#include <string>
	using namespace std;

	class Color {
		public:
			int key;

			Color(string name) {
				this->name = name;
				this->disease_cubes = 24;
				this->cured = 0;
			}

			bool isCured() { return cured >= 1; }
			bool isEradicated() { return cured == 2; }
			int getDiseaseCubes() { return disease_cubes; }
			string getName() { return name; }

			void cure() {
				if (cured == 0) {
					setCured(1);
				}
				if (disease_cubes == 24) {
					setCured(2);
				}
			}

			void eradicate() {
				if (cured == 1 && disease_cubes == 24) {
					setCured(2);
				}
			}

			int takeCubes(int num) {
				if (num <= 0) {
					throw invalid_argument("Invalid number of cubes to take. Must take at least 1 cube");
				}
				if (num > disease_cubes) {
					throw invalid_argument("Not enough cubes to take from " + name + ". Game Lost.");
				}
				disease_cubes -= num;
				return num;
			}

			int returnCubes(int num) {
				if (num <= 0) {
					throw invalid_argument("Invalid number of cubes to return. Must return at least 1 cube");
				}
				if (disease_cubes + num > 24) {
					throw invalid_argument("Too many cubes to return to " + name);
				}

				disease_cubes += num;


				return num;
			}

		private:
			string name;
			int disease_cubes; // Number of disease cubes left for this color
			int cured; // 0 for uncured, 1 for cured, 2 for eradicated

			void setCured(int cured) {
				if (cured < 0 || cured > 2) {
					throw invalid_argument("Invalid value cured (" + cured + ") for setCured().");
				}
				this->cured = cured;
			}
	};
}
